From 4c659150f8daea86bef5531b7c15f4385f971db9 Mon Sep 17 00:00:00 2001
From: Jingyu Zhou <b02241@freescale.com>
Date: Fri, 1 Aug 2008 10:47:50 +0800
Subject: [PATCH] ENGR00086183 Enable alsa mixer all output control when no playback active

Enable alsa mixer all output control when no playback active

Signed-off-by: Zhou Jingyu <Jingyu.Zhou@freescale.com>
---
 sound/arm/mxc-alsa-pmic.c |   16 ++++++++++++++++
 1 files changed, 16 insertions(+), 0 deletions(-)

diff --git a/sound/arm/mxc-alsa-pmic.c b/sound/arm/mxc-alsa-pmic.c
index 392dab2..29b087c 100644
--- a/sound/arm/mxc-alsa-pmic.c
+++ b/sound/arm/mxc-alsa-pmic.c
@@ -1375,10 +1375,16 @@ int set_mixer_output_device(PMIC_AUDIO_HANDLE handle, OUTPUT_SOURCE src,
 		if (audio_mixer_control.output_device & SOUND_MASK_PHONEOUT) {
 			audio_mixer_control.output_active[OP_EARPIECE] = 1;
 			pmic_audio_output_set_port(handle, MONO_SPEAKER);
+		} else {
+			audio_mixer_control.output_active[OP_EARPIECE] = 0;
+			pmic_audio_output_clear_port(handle, MONO_SPEAKER);
 		}
 		if (audio_mixer_control.output_device & SOUND_MASK_SPEAKER) {
 			audio_mixer_control.output_active[OP_HANDSFREE] = 1;
 			pmic_audio_output_set_port(handle, MONO_LOUDSPEAKER);
+		} else {
+			audio_mixer_control.output_active[OP_HANDSFREE] = 0;
+			pmic_audio_output_clear_port(handle, MONO_LOUDSPEAKER);
 		}
 		if (audio_mixer_control.output_device & SOUND_MASK_VOLUME) {
 			audio_mixer_control.output_active[OP_HEADSET] = 1;
@@ -1392,12 +1398,22 @@ int set_mixer_output_device(PMIC_AUDIO_HANDLE handle, OUTPUT_SOURCE src,
 				/*This is a temporary workaround which should be removed later */
 			}
 
+		} else {
+			audio_mixer_control.output_active[OP_HEADSET] = 0;
+			pmic_audio_output_clear_port(handle,
+						     STEREO_HEADSET_LEFT |
+						     STEREO_HEADSET_RIGHT);
 		}
 		if (audio_mixer_control.output_device & SOUND_MASK_PCM) {
 			audio_mixer_control.output_active[OP_LINEOUT] = 1;
 			pmic_audio_output_set_port(handle,
 						   STEREO_OUT_LEFT |
 						   STEREO_OUT_RIGHT);
+		} else {
+			audio_mixer_control.output_active[OP_LINEOUT] = 0;
+			pmic_audio_output_clear_port(handle,
+						     STEREO_OUT_LEFT |
+						     STEREO_OUT_RIGHT);
 		}
 	} else {
 		switch (dev) {
-- 
1.5.4.4

